Mathlib/Order/SuccPred/InverseSystem.lean

```diff
@@ -0,0 +1,294 @@
+which we would like to "glue" up to a bijection `F i ≃ ∀ l : Iio i, X l`. We denote
+`∀ l : Iio i, X l` by `PiLT X i`, and they form an inverse system just like the `F i`.
+Observe that at a limit node `i`, `PiLT X i` is actually the inverse limit of `PiLT X j` over
+all `j < i` (`piLTLim`). If the family of bijections `F j ≃ PiLT X j` is natural (`IsNatEquiv`),
+we immediately obtain a bijection between the limits `limit F i ≃ PiLT X i` (`invLimEquiv`),
+and we just need an additional bijection `F i ≃ limit F i` to obtain the desired
+extension `F i ≃ PiLT X i` to the limit node `i`. (We do have such a bijection, for example, when
+we consider a directed system of algebraic structures (say fields) `K i`, and `F` is
+the inverse system of homomorphisms `K i ⟶ K` into a specific field `K`.)
+
+Now our task reduces to recursive construction of a *natural* family of bijections for each `i`.
+We can prove by that a natural family over all `l ≤ i` (`Iic i`) extends to a natural family over
+`Iic i⁺`, but at a limit node, recursion stops working: we have natural families over all `Iic j`
+for each `j < i`, but we need to know that they glue together to form a natural family over all
+`l < i` (`Iio i`). This intricacy did not occur to the author when he thought he had a proof and
+set out to formalize it. Fortunately he was able to figure out an additional `compat` condition
+(compatibility with the bijections `F i⁺ ≃ F i × X i` in the `X` component) that guarantees
+uniqueness (`unique_pEquivOn`) and hence gluability (well-definedness): see `pEquivOnGlue`.
+Instead of just a family of natural families, we actually construct a family of the stronger
+`PEquivOn`s that bundles the `compat` condition, in order for the inductive argument to work.
+
+It is possible to circumvent the introduction of the `compat` condition using Zorn's lemma;
+if there is a chain of natural families (i.e. for any two families in the chain, one is an
+extension of the other) over lowersets (which are all of the form `Iic`, `Iio`, or `univ`),
+we can clearly take the union to get a natural family that extend them all. If a maximal
+natural family has domain `Iic i` or `Iio i` (`i` a limit), we already know how to extend it
+one step further to `Iic i⁺` or `Iic i` respectively, so it must be the case that the domain
+is everything. However, the author chose the `compat` approach in the end because it constructs
+the distinguished bijection that is compatible with the projections to all `X i`.
+
+-/
+
+open Order Set
+
+variable {ι : Type*} [Preorder ι] {F X : ι → Type*} ⦃i j : ι⦄ (h : i ≤ j)
+  (f : ∀ ⦃i j : ι⦄, i ≤ j → F j → F i)
+
+/-- A directed system indexed by a preorder is a contravariant functor from the preorder
+to another category. It is dual to `DirectedSystem`. -/
+class InverseSystem : Prop where
+  map_self ⦃i⦄ (x : F i) : f le_rfl x = x
+  map_map ⦃k j i⦄ (hkj : k ≤ j) (hji : j ≤ i) (x : F i) : f hkj (f hji x) = f (hkj.trans hji) x
```

Comment:
Could you move the definition of `DirectedSystem` to this file?

Reply:
and then import this file into Mathlib.Algebra.DirectLimit? That file currently probably doesn't require/import as much order theory as this file. Let's see whether we get a large-import alert ...

Reply:
> and then import this file into Mathlib.Algebra.DirectLimit?

Yes exactly

> That file currently probably doesn't require/import as much order theory as this file. Let's see whether we get a large-import alert ...

I believe it should be fine. You are not importing a huge lot here, and Algebra.DirectLimit is supposed to be quite advanced

Reply:
You're right, it appears that Algebra.DirectLimit doesn't get any new imports!